File: Day 2 - Easy (Level 1)/RabbitVoting/RabbitVoting.cs
// BEGIN CUT HERE

// END CUT HERE
using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;

public class RabbitVoting {
    public string getWinner(string[] names, string[] votes) {
        string res;
        return res;
    }

// BEGIN CUT HERE
    public static void Main(String[] args) {
        try  {
            eq(0,(new RabbitVoting()).getWinner(new string[] { "Alice", "Bill", "Carol", "Dick" }, new string[] { "Bill", "Dick", "Alice", "Alice" }),"Alice");
            eq(1,(new RabbitVoting()).getWinner(new string[] { "Alice", "Bill", "Carol", "Dick" }, new string[] { "Carol", "Carol", "Bill", "Bill" }),"");
            eq(2,(new RabbitVoting()).getWinner(new string[] { "Alice", "Bill", "Carol", "Dick" }, new string[] { "Alice", "Alice", "Bill", "Bill" }),"Bill");
            eq(3,(new RabbitVoting()).getWinner(new string[] { "Alice", "Bill" }, new string[] { "Alice", "Bill" }),"");
            eq(4,(new RabbitVoting()).getWinner(new string[] { "WhiteRabbit", "whiterabbit", "whiteRabbit", "Whiterabbit" }, new string[] { "whiteRabbit", "whiteRabbit", "whiteRabbit", "WhiteRabbit" }),"whiteRabbit");
        } 
        catch( Exception exx)  {
            System.Console.WriteLine(exx);
            System.Console.WriteLine( exx.StackTrace);
        }
    }
    private static void eq( int n, object have, object need) {
        if( eq( have, need ) ) {
            Console.WriteLine( "Case "+n+" passed." );
        } else {
            Console.Write( "Case "+n+" failed: expected " );
            print( need );
            Console.Write( ", received " );
            print( have );
            Console.WriteLine();
        }
    }
    private static void eq( int n, Array have, Array need) {
        if( have == null || have.Length != need.Length ) {
            Console.WriteLine("Case "+n+" failed: returned "+have.Length+" elements; expected "+need.Length+" elements.");
            print( have );
            print( need );
            return;
        }
        for( int i= 0; i < have.Length; i++ ) {
            if( ! eq( have.GetValue(i), need.GetValue(i) ) ) {
                Console.WriteLine( "Case "+n+" failed. Expected and returned array differ in position "+i );
                print( have );
                print( need );
                return;
            }
        }
        Console.WriteLine("Case "+n+" passed.");
    }
    private static bool eq( object a, object b ) {
        if ( a is double && b is double ) {
            return Math.Abs((double)a-(double)b) < 1E-9;
        } else {
            return a!=null && b!=null && a.Equals(b);
        }
    }
    private static void print( object a ) {
        if ( a is string ) {
            Console.Write("\"{0}\"", a);
        } else if ( a is long ) {
            Console.Write("{0}L", a);
        } else {
            Console.Write(a);
        }
    }
    private static void print( Array a ) {
        if ( a == null) {
            Console.WriteLine("<NULL>");
        }
        Console.Write('{');
        for ( int i= 0; i < a.Length; i++ ) {
            print( a.GetValue(i) );
            if( i != a.Length-1 ) {
                Console.Write(", ");
            }
        }
        Console.WriteLine( '}' );
    }
// END CUT HERE
}

File: Day 2 - Easy (Level 1)/RabbitVoting/RabbitVoting.cpp
// BEGIN CUT HERE

// END CUT HERE
#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <queue>
#include <set>
#include <map>
#include <cstdio>
#include <cstdlib>
#include <cctype>
#include <cmath>
using namespace std;

vector<string> split( const string& s, const string& delim =" " ) {
    vector<string> res;
    string t;
    for ( int i = 0 ; i != s.size() ; i++ ) {
        if ( delim.find( s[i] ) != string::npos ) {
            if ( !t.empty() ) {
                res.push_back( t );
                t = "";
            }
        } else {
            t += s[i];
        }
    }
    if ( !t.empty() ) {
        res.push_back(t);
    }
    return res;
}

vector<int> splitInt( const string& s, const string& delim =" " ) {
    vector<string> tok = split( s, delim );
    vector<int> res;
    for ( int i = 0 ; i != tok.size(); i++ )
        res.push_back( atoi( tok[i].c_str() ) );
    return res;
}

// BEGIN CUT HERE
#define ARRSIZE(x) (sizeof(x)/sizeof(x[0]))

template<typename T> void print( T a ) {
    cerr << a;
}
static void print( long long a ) {
    cerr << a << "L";
}
static void print( string a ) {
    cerr << '"' << a << '"';
}
template<typename T> void print( vector<T> a ) {
    cerr << "{";
    for ( int i = 0 ; i != a.size() ; i++ ) {
        if ( i != 0 ) cerr << ", ";
        print( a[i] );
    }
    cerr << "}" << endl;
}
template<typename T> void eq( int n, T have, T need ) {
    if ( have == need ) {
        cerr << "Case " << n << " passed." << endl;
    } else {
        cerr << "Case " << n << " failed: expected ";
        print( need );
        cerr << " received ";
        print( have );
        cerr << "." << endl;
    }
}
template<typename T> void eq( int n, vector<T> have, vector<T> need ) {
    if( have.size() != need.size() ) {
        cerr << "Case " << n << " failed: returned " << have.size() << " elements; expected " << need.size() << " elements.";
        print( have );
        print( need );
        return;
    }
    for( int i= 0; i < have.size(); i++ ) {
        if( have[i] != need[i] ) {
            cerr << "Case " << n << " failed. Expected and returned array differ in position " << i << ".";
            print( have );
            print( need );
            return;
        }
    }
    cerr << "Case " << n << " passed." << endl;
}
static void eq( int n, string have, string need ) {
    if ( have == need ) {
        cerr << "Case " << n << " passed." << endl;
    } else {
        cerr << "Case " << n << " failed: expected ";
        print( need );
        cerr << " received ";
        print( have );
        cerr << "." << endl;
    }
}
// END CUT HERE
class RabbitVoting {
public:
    string getWinner(vector <string> names, vector <string> votes) {
        string res;
        return res;
    }

};
// BEGIN CUT HERE
void main( int argc, char* argv[] ) {
    {
        string namesARRAY[] = { "Alice", "Bill", "Carol", "Dick" };
        vector <string> names( namesARRAY, namesARRAY+ARRSIZE(namesARRAY) );
        string votesARRAY[] = { "Bill", "Dick", "Alice", "Alice" };
        vector <string> votes( votesARRAY, votesARRAY+ARRSIZE(votesARRAY) );
        RabbitVoting theObject;
        eq(0, theObject.getWinner(names, votes),"Alice");
    }
    {
        string namesARRAY[] = { "Alice", "Bill", "Carol", "Dick" };
        vector <string> names( namesARRAY, namesARRAY+ARRSIZE(namesARRAY) );
        string votesARRAY[] = { "Carol", "Carol", "Bill", "Bill" };
        vector <string> votes( votesARRAY, votesARRAY+ARRSIZE(votesARRAY) );
        RabbitVoting theObject;
        eq(1, theObject.getWinner(names, votes),"");
    }
    {
        string namesARRAY[] = { "Alice", "Bill", "Carol", "Dick" };
        vector <string> names( namesARRAY, namesARRAY+ARRSIZE(namesARRAY) );
        string votesARRAY[] = { "Alice", "Alice", "Bill", "Bill" };
        vector <string> votes( votesARRAY, votesARRAY+ARRSIZE(votesARRAY) );
        RabbitVoting theObject;
        eq(2, theObject.getWinner(names, votes),"Bill");
    }
    {
        string namesARRAY[] = { "Alice", "Bill" };
        vector <string> names( namesARRAY, namesARRAY+ARRSIZE(namesARRAY) );
        string votesARRAY[] = { "Alice", "Bill" };
        vector <string> votes( votesARRAY, votesARRAY+ARRSIZE(votesARRAY) );
        RabbitVoting theObject;
        eq(3, theObject.getWinner(names, votes),"");
    }
    {
        string namesARRAY[] = { "WhiteRabbit", "whiterabbit", "whiteRabbit", "Whiterabbit" };
        vector <string> names( namesARRAY, namesARRAY+ARRSIZE(namesARRAY) );
        string votesARRAY[] = { "whiteRabbit", "whiteRabbit", "whiteRabbit", "WhiteRabbit" };
        vector <string> votes( votesARRAY, votesARRAY+ARRSIZE(votesARRAY) );
        RabbitVoting theObject;
        eq(4, theObject.getWinner(names, votes),"whiteRabbit");
    }
}
// END CUT HERE
